src/services/game-logic/GameEngine.ts

```diff
@@ -12,22 +12,52 @@ class GameEngine {
 
     getLegalMoves = (square: Square): Square[] => {
         const piece = this.board.getPiece(square);
-        let pieceMoves = piece?.getPossibleMoves() ?? [];
-        return pieceMoves
-            .filter(this.isOnBoard)
-            .filter((square) => !this.isPawnAndBlocked(square, piece))
-            .filter((square) => !this.isOccupiedBySameColor(square, piece));
+        let legalMoves = piece?.getPossibleMoves().filter(this.isOnBoard) ?? [];
+        if (piece?.name === PieceNames.KING)
+            legalMoves = legalMoves.filter((square) => !this.isCastlingIllegal(square, piece));
```

Comment:
podwojne zaprzeczenie jest często ciężkie do zrozumienia. odpowiednio zmieniłbym ciało metody żeby można było ją nazywać isCastlingLegal i wtedy tu nie trzeba robić zaprzeczenia :)

Reply:
można tez zmienić nazwę parametru lambdy w filter np na move albo legalMove bo nazywa się tak samo jak parametr metody

Reply:
dzięki za uwagi!, poprawiłam te nazwy, magic numbers i dodałam nawiasy klamrowe. Jeśli chodzi o funkcję isCastlingIllegal, to choć sama nie lubię podwójnych zaprzeczeń, to uznałam, że tutaj funkcja musi mieć taką konstrukcję, bo trzeba obsłużyć ruchy, które nie są roszadą (3 sytuacje - nie jest roszadą, jest prawidłową roszadą i jest nielegalną roszadą). Gdyby próbować to odwrócić to funkcja nazywałaby się isNotCastlingOrLegalCastling

Reply:
@Nebrakada rozprawiłam się z tą funkcją podwójnym zaprzeczeniem ;( dasz proszę approve, żebym mogła zmergować? dzięki!